lib/ansible/vars/manager.py

```diff
@@ -269,27 +284,31 @@ def plugins_by_groups():
                     display.warning('Ignoring unknown variable precedence entry: %s' % (entry))
 
             # host vars, from inventory, inventory adjacent and play adjacent via plugins
-            all_vars = _combine_and_track(all_vars, host.get_vars(), "host vars for '%s'" % host)
+            host_vars = host.get_vars()
+            for i in HOST_INTERNAL:
+                # keep internal vars to avoid overwrites
+                if i in host_vars:
+                    internal[i] = host_vars.pop(i)
+            all_vars = _combine_and_track(all_vars, host_vars, "host vars for '%s'" % host)
             all_vars = _combine_and_track(all_vars, _plugins_inventory([host]), "inventory host_vars for '%s'" % host)
             all_vars = _combine_and_track(all_vars, _plugins_play([host]), "playbook host_vars for '%s'" % host)
 
             # finally, the facts caches for this host, if it exists
-            # TODO: cleaning of facts should eventually become part of taskresults instead of vars
             try:
                 try:
                     facts = self._fact_cache.get(host.name)
                 except KeyError:
                     facts = {}
 
-                all_vars |= namespace_facts(facts)
+                ns = namespace_facts(facts)
+                internal['ansible_facts'] = ns.pop('ansible_facts', {})
+                internal['ansible_local'] = internal['ansible_facts'].get('ansible_local', {})  # always have 'local'
```

Comment:
There's a test relying on being able to set ansible_local, pretty sure this is causing the CI failure.

Reply:
I know, still debating if the test is correct usage or not, the same for `ansible_facts` itself,  there are examples online of plays just setting it.